tests/foreman/longrun/test_oscap.py

```diff
@@ -415,50 +530,37 @@ def test_positive_oscap_run_via_ansible_bz_1814988(
             'organizations': module_org.name,
         }
     )
-    with Broker(nick='rhel7', host_class=ContentHost, deploy_flavor=settings.flavors.default) as vm:
-        result = vm.register(module_org, None, ak_name['rhel7'], target_sat)
-        assert result.status == 0, f'Failed to register host: {result.stderr}'
-        vm.create_custom_repos(rhel7=settings.repos.rhel7_os)
-        # Harden the rhel7 client with DISA STIG security policy
-        vm.run('yum install -y scap-security-guide')
-        vm.run(
-            'oscap xccdf eval --remediate --profile xccdf_org.ssgproject.content_profile_stig '
-            '--fetch-remote-resources --results-arf results.xml '
-            '/usr/share/xml/scap/ssg/content/ssg-rhel7-ds.xml',
-        )
-        target_sat.cli.Host.update(
-            {
-                'name': vm.hostname.lower(),
-                'lifecycle-environment': lifecycle_env.name,
-                'content-view': content_view.name,
-                'hostgroup': hgrp_name,
-                'openscap-proxy-id': default_proxy,
-                'organization': module_org.name,
-                'ansible-role-ids': role_id,
-            }
-        )
-        job_id = target_sat.cli.Host.ansible_roles_play({'name': vm.hostname.lower()})[0].get('id')
-        target_sat.wait_for_tasks(
-            f'resource_type = JobInvocation and resource_id = {job_id} and action ~ "hosts job"'
+    target_sat.cli.Host.update(
+        {
+            'name': contenthost.hostname.lower(),
+            'lifecycle-environment': lifecycle_env.name,
+            'content-view': cv_name[distro],
+            'hostgroup': hgrp_name,
+            'openscap-proxy-id': default_proxy,
+            'organization': module_org.name,
+            'ansible-role-ids': role_id,
+        }
+    )
```

Comment:
Shouldn't this already be done as part of the registration? At least the organization is already mandatory for registration and the hostgroup can be set there. What you can then do is set the remaining options on the hostgroup (that you create on line 508) instead and let the host inherit them.

In theory you could then use the same hostgroup for all RHEL versions, but that's probably not the best idea in terms of test design.

Still, creating entities correctly once saves an update and makes the test easier to follow.

Reply:
It is clumsy but the reason is that the content is sourced from the clients along the lines of https://docs.redhat.com/en/documentation/red_hat_satellite/6.15/html/managing_security_compliance/configuring_scap_contents_security-compliance#getting-supported-scap-contents-for-rhel_security-compliance, so it doesn't exist on Satellite at registration time.

Reply:
But none of the data relies on that. Only to create the actual SCAP policy. I didn't test this, but isn't this a more logical way of writing the test:
```python
@pytest.mark.rhel_ver_list([7, 8, 9])
@pytest.mark.tier4
def test_positive_oscap_run_via_ansible_bz_1814988(
    module_org, default_proxy, lifecycle_env, target_sat, rex_contenthost
):
    """End-to-End Oscap run with DISA remediation

    :id: 375f8f08-9299-4d16-91f9-9426eeecb9c5

    :parametrized: yes

    :customerscenario: true

    :setup: scap content, scap policy, host group

    :steps:

        1. Import Ansible role theforeman.foreman_scap_client
        2. Import Ansible Variables needed for the role
        3. Create a hostgroup
        4. Provision a host using the hostgroup
        5. Harden the host by remediating it with DISA STIG security policy
        6. Create a valid scap content
        7. Create a scap policy associated with the hostgroup and ansible as deploy option
        8. Play roles for the host

    :expectedresults: REX job should be success and ARF report should be sent to satellite

    :BZ: 1814988
    """
    contenthost = rex_contenthost
    os_version = contenthost.os_version.major
    distro = f'rhel{os_version}'

    # Ensure Ansible role
    target_sat.cli.Ansible.roles_import({'proxy-id': default_proxy})
    target_sat.cli.Ansible.variables_import({'proxy-id': default_proxy})
    role_id = target_sat.cli.Ansible.roles_list({'search': 'foreman_scap_client'})[0].get('id')

    # Create a hostgroup
    hostgroup = target_sat.cli_factory.hostgroup(
        {
            'content-source-id': default_proxy,
            'name': gen_string('alpha'),
            'organizations': module_org.name,
            'lifecycle-environment': lifecycle_env.name,
            'content-view': cv_name[distro],
            'openscap-proxy-id': default_proxy,
            'ansible-role-ids': role_id,
        }
    )

    # Create a host
    result = contenthost.register(
        org=module_org,
        loc=None,
        activation_keys=ak_name[distro],
        target=target_sat,
        ignore_subman_errors=True,
        force=True,
        insecure=True,
        hostgroup=hostgroup,
    )
    assert result.status == 0, f'Failed to register host: {result.stderr}'
    rhel_repo = rhel_repos[distro]
    if distro == 'rhel7':
        contenthost.create_custom_repos(distro=rhel_repo)
    else:
        contenthost.create_custom_repos(**rhel_repo)

    # Harden the client with DISA STIG security policy
    contenthost.run('yum install -y scap-security-guide')
    contenthost.run(
        'oscap xccdf eval --remediate --profile xccdf_org.ssgproject.content_profile_stig '
        '--fetch-remote-resources --results-arf results.xml '
        f'/usr/share/xml/scap/ssg/content/ssg-{distro}-ds.xml',
    )

    # Create SCAP content
    content = find_content_to_update(target_sat, module_org, distro, contenthost)
    update_scap_content(module_org, target_sat, content)

    # Create oscap_policy.
    scap_id, scap_profile_id = fetch_scap_and_profile_id(target_sat, content, profiles[distro])
    target_sat.cli_factory.make_scap_policy(
        {
            'scap-content-id': scap_id,
            'hostgroups': hostgroup.name,
            'deploy-by': 'ansible',
            'name': gen_string('alpha'),
            'period': OSCAP_PERIOD['weekly'].lower(),
            'scap-content-profile-id': scap_profile_id,
            'weekday': OSCAP_WEEKDAY['friday'].lower(),
            'organizations': module_org.name,
        }
    )

    # Apply policy
    job_id = target_sat.cli.Host.ansible_roles_play({'name': contenthost.hostname.lower()})[0].get(
        'id'
    )
    target_sat.wait_for_tasks(
        f'resource_type = JobInvocation and resource_id = {job_id} and action ~ "hosts job"'
    )
    try:
        result = target_sat.cli.JobInvocation.info({'id': job_id})['success']
        assert result == '1'
    except AssertionError as err:
        output = ' '.join(
            target_sat.cli.JobInvocation.get_output({'id': job_id, 'host': contenthost.hostname})
        )
        result = f'host output: {output}'
        raise AssertionError(result) from err
    result = contenthost.run('grep profile /etc/foreman_scap_client/config.yaml')
    assert result.status == 0
    contenthost.execute_foreman_scap_client()
    result = target_sat.cli.Arfreport.list({'search': f'host={contenthost.hostname.lower()}'})
    assert result is not None
```

I did question why you imported Ansible variables, because those don't appear to be used in the tests.

You may notice I removed some helper variables that were used only once and moved the order of steps around. To make it even clearer, each section now has a comment above it that corresponds to a step in the description.

One thing I noticed is that some steps look more like fixtures.

This is way out of scope for this PR, but there's a huge overlap between two tests here. One does exactly the same as the other, except it also runs the DISA step. Something like https://smarie.github.io/python-pytest-steps/ could help with separating those out and allowing reuse. As for the various RHELs: you provide a lot of configuration but I wonder if https://smarie.github.io/python-pytest-cases/ wouldn't be a better way describe those case.

Reply:
@ekohl hi, mind revisiting?I made the suggested change to remove the host update everywhere where it occured. I'm not certain about the future of the DISA remediation test, as it covers a scenario of not yet resolved bug, it might as well be removed. But I take note about pytest-steps for the futrue, thanks.